engine/src/juliabox/plugins/compute_gce/impl_gce.py

```diff
@@ -0,0 +1,650 @@
+        return ins.get(project=CompGCE.INSTALL_ID, zone=CompGCE._zone(),
+                       instance=instance_name).execute()
+
+    @staticmethod
+    def _uptime_minutes(instance_name=None):
+        attrs = CompGCE._instance_attrs(instance_name)
+        lt = parse_iso_time(attrs["creationTimestamp"])
+        nt = datetime.datetime.now(pytz.utc)
+        uptime = nt - lt
+        minutes = int(uptime.total_seconds()/60)
+        return minutes
+
+    @staticmethod
+    @retry_on_bsl
+    def _get_instances(gname, only_running=False):
+        conn = CompGCE._connect_gce().instanceGroups()
+        flag = 'RUNNING' if only_running else 'ALL'
+        data = conn.listInstances(project=CompGCE.INSTALL_ID, zone=CompGCE._zone(),
+                                  body={'instanceState': flag},
+                                  instanceGroup=gname).execute()
+        ret = []
+        for inst in data.get('items', []):
+            ret.append(inst['instance'].split('/')[-1])
+        return ret
+
+    @staticmethod
+    def get_all_instances(gname=None):
+        if gname is None:
+            gname = CompGCE.AUTOSCALE_GROUP
+        if gname is None:
+            return [CompGCE.get_instance_id()]
+        instances = CompGCE._get_instances(gname, only_running=True)
+        if len(instances) == 0:
+            return [CompGCE.get_instance_id()]
+        return instances
+
+    @staticmethod
+    @retry_on_bsl
+    def _increment_num_instances():
+        conn = CompGCE._connect_gce().instanceGroupManagers()
+        curr = conn.get(project=CompGCE.INSTALL_ID, zone=CompGCE._zone(),
+                        instanceGroupManager=CompGCE.AUTOSCALE_GROUP).execute()['targetSize']
+        return conn.resize(project=CompGCE.INSTALL_ID, zone=CompGCE._zone(),
+                           instanceGroupManager=CompGCE.AUTOSCALE_GROUP,
+                           size=curr + 1).execute()
+
+    @staticmethod
+    def _get_last_scale_up_time():
+        plugin = JBPluginDB.jbox_get_plugin(JBPluginDB.JBP_DB)
+        table = plugin.table_open('scale_up_time')
+        return [x for x in plugin.record_scan(table)][0]['scale_up_time']
+
+    @staticmethod
+    def _set_last_scale_up_time(t):
+        plugin = JBPluginDB.jbox_get_plugin(JBPluginDB.JBP_DB)
+        conn = plugin.conn()
+        c = conn.cursor()
+        c.execute('UPDATE scale_up_time set scale_up_time = %d' % t)
+        c.close()
+        conn.commit()
+
+    @staticmethod
+    def _add_instance():
+        try:
+            # Execute policy only after a reasonable wait period to let a new machine boot up.
+            # This will prevent thrashing AWS APIs and triggering AWS throttling.
+            # Cooldown policy can also apply after that.
+            now = int(time.time())
+            last_time = CompGCE._get_last_scale_up_time()
+            if now > last_time + CompGCE.SCALE_UP_INTERVAL:
+                CompGCE._set_last_scale_up_time(now)
```

Comment:
This still has a race condition, where multiple instances can call `_get_last_scale_up_time` at the same time, find it not recent and proceed to call `_set_last_scale_up_time` and `_increment_num_instances`.

One way to handle it is to add a clause in `_set_last_scale_up_time` to check for `scale_up_time` to be same as last observed.